### ISSUE TITLE
KeePassXC DB in a folder with cyrillic letter in a path

### ISSUE THREAD

**Issue by alexus-m** [asbru-cm/asbru-cm#859]
**Describe the bug**
If a KeePassXC kdbx file stored in a folder with cyrillic letter in a path to the file, then Asbru-CM "Edit Connection" window can successfully open and unlock the configured kdbx file, then locate there a requested record and assign it to the connection properties. But during connection attempt it cannot find the saved entry in kdbx with an error message in connection terminal window:
"ERROR Unable to find entry <username|/Test//192.168.0.1> in the configured KeePass database /home/mayorov/Документы/asbru-cm.kdbx]."
If the same kdbx file moved to a folder with latin letters only, e.g. /home/mayorov/Documenty/asbru-cm.kdbx, then it works as expected: can open and extract the login data from kdbx file and successfully connect to a ssh server.

**Expected behavior**
Asbru-CM should:
1. successfully open a kdbx file during connection attempt if a path to the kdbx has localized letters, e.g. cyrillic, as it can open the kdbx file during configuration of the connection.
2. correctly show a path with cyrillic letters. (take a look on a screenshot of unlock window)

**Screenshots**
![KeePassXC_Unlock](https://user-images.githubusercontent.com/20230747/139586347-3bebe905-a9d8-4b8d-acbe-86f4e396241b.png)

**Environment (please complete the following information):**
 - OS: Fedora 35
 - Ásbrú Version: loki (asbru-cm-6.4.0~1635596963-1.fc34.noarch)

**Comment by hanspr** [comment]
While the fix is merged if you do not mind do some editing, to keep working.

1. Exit asbru
2. With a text editor
3. Open file ~/.config/asbru/asbur.yml
4. Locate the name of your keepass file. You will see that has scrambled characters.
5. Edit the full path.
6. Save.
7. Start asbru again

The path will be correct, and should work.

**Comment by alexus-m** [comment]
Hi @hanspr !
Thank you for your comment and patch!
  I did a try to manually edit asbru.yml file without applied proposed patch but it does not work: 'database' value is overwritten in config file every time I started asbru and quit. (All encoded passwords are rewritten as well on every start. - May be it is because I changed Crypt::CBC salt from predefined value to the random salt=1. But with predefined one a has another issue #858.)

  Regarding to the proposed patch:
I don't know does it need to decode 'pathcli' (has no glue to a possible scenario) but 'keyfile' could be required to open kdbx and it can be stored in a folder with localized names. Therefore it's required to convert 'keyfile' as well. I tested your patch without decode 'keyfile' - it does not work if key file is stored in a folder with cyrillic letters. After addition of decode operation to the 'keyfile' then path in asbru.yml file get unscrambled and start to work.
So please add decode to a line with
'$hash{keyfile} ='

**Comment by hanspr** [comment]
Hi @alexus-m ,

The decoding is over the paths not over the content of the files, if now fails to find the  correct location of the database file, will have the same problem with the key path. The problem is that at the moment of writing the content to the file the Yalm library encodes everything to utf8, and those values are already in utf8, so they get double encoded.

Regarding the edition, then

1. Exit asbru
2. remove file ~/.config/asbur/asbru.nfreeze
2. With a text editor
3. Open file ~/.config/asbru/asbru.yml
4. Locate the name of your keepass file. You will see that has scrambled characters.
5. Edit the full path.
6. Save.
7. Start asbru again

asbru.nfreeze is some binary copy of the yalm file to cache it and load faster, it is removed and rebuild when you change the settings, the application is not aware or your changes so it writes it back.

**Comment by alexus-m** [comment]
Hi @hanspr !
Thanks a lot for you support!
  Without your patch: if I have removed asbru.nfreeze and replaced in asbru.yml scrambled path with readable cyrillic text, then asbru-cm works through some starts as expected even on not patched PACKeePass.pm without further interventions until I change file on KeePass integration page.
  I applied your patch to decode of paths. - Kdbx manipulation works as expected but path to Key file is still affected. After addition decode to 'keyfile' everything started to work like a charm.

$ diff /usr/share/asbru-cm/lib/PACKeePass.pm.fixed /usr/share/asbru-cm/lib/PACKeePass.pm.orig
301,303c301,303
<     $hash{database} = decode('utf8',$$self{frame}{'fcbKeePassFile'}->get_filename());
<     $hash{pathcli} = decode('utf8',$$self{frame}{'fcbCliFile'}->get_filename());
<     $hash{keyfile} = **decode**('utf8',$$self{frame}{'fcbKeePassKeyFile'}->get_filename());
---
>     $hash{database} = $$self{frame}{'fcbKeePassFile'}->get_filename();
>     $hash{pathcli} = $$self{frame}{'fcbCliFile'}->get_filename();
>     $hash{keyfile} = $$self{frame}{'fcbKeePassKeyFile'}->get_filename();
707c707
<             $$self{cfg}{pathcli} = decode('utf8',$fc->get_filename());
---
>             $$self{cfg}{pathcli} = $fc->get_filename();

**Comment by hanspr** [comment]
@alexus-m 

Glad it worked out, and thanks for pointing out that I missed the decoding of the key file path, had fixed this in the PR

**Comment by alexus-m** [comment]
The graphical prompt to unblock KeePass dbx file is corrected after patch as well.
Thanks!
![KeePassXC_Unlock fixed](https://user-images.githubusercontent.com/20230747/139946595-67642afc-e522-4941-8b07-d25741fc3495.png)

**Comment by alexus-m** [comment]
Just to mention:
I found description of the mentioned trick with deletion of asbru.nfreeze in case of manual edit asbru.yml in "Tips and Tricks" chapter of Wiki.
https://github.com/asbru-cm/asbru-cm/wiki/Tips-and-Tricks#my-changes-to-configasbruasbruyml-are-ignored-what-is-wrong-

**Comment by gfrenoy** [comment]
@alexus-m Can you confirm it's now working as expected on the `loki` branch ?

**Comment by alexus-m** [comment]
Dear team,
I'm happy to confirm that access to kdbx and key files with cyrillic letters in paths works as expected in loki now.
To test access to cli I've copied system wide keepassxc-cli into my 'Документы' folder and pointed asbru to the file. Asbru accepted the path, detected version of cli and worked with the copy of cli file even after asbru restart. I'd assume that access to cli should work well even if a user's custom version placed in a folder with localized path, e.g. cyrillic at least.
Thank you a lot for your rapid and strong support!
